### PR TITLE
RESOURCE-370 AWS resource_ids - EC2

### PR DESCRIPTION
**Pull request by soumyo13**
Signed-off-by: Soumyodeep Karmakar <soumyo.k13@gmail.com>

### Description

The following resources are added with the resource_id.

                Amazon EC2 Auto Scaling:
                    aws_autoscaling_scaling_policy
                    aws_launch_configuration
                
                Application Auto Scaling:
                    aws_application_autoscaling_scalable_target
                    aws_application_autoscaling_scaling_policy
                
                EC2:
                    aws_ami
                    aws_dhcp_options
                    aws_ebs_snapshot
                    aws_ebs_volume
                    aws_ec2_capacity_reservation
                    aws_ec2_carrier_gateway
                    aws_ec2_client_vpn_authorization_rule
                    aws_ec2_client_vpn_endpoint
                    aws_ec2_client_vpn_route
                    aws_ec2_client_vpn_target_network_association
                    aws_ec2_customer_gateway
                    aws_ec2_dhcp_option
                    aws_ec2_egress_only_internet_gateway
                    aws_ec2_eip
                    aws_ec2_eip_association
                    aws_ec2_fleet
                    aws_ec2_host
                    aws_ec2_instance
                    aws_ec2_internet_gateway
                    aws_ec2_launch_template
                    aws_ec2_network_insights_analysis
                    aws_ec2_network_insights_path
                    aws_ec2_network_interface
                    aws_ec2_network_interface_attachment
                    aws_ec2_placement_group
                    aws_ec2_prefix_list
                    aws_ec2_spot_fleet
                    aws_ec2_traffic_mirror_filter
                    aws_ec2_traffic_mirror_session
                    aws_ec2_traffic_mirror_target
                    aws_ec2_transit_gateway_attachment
                    aws_ec2_transit_gateway_route_table
                    aws_ec2_transit_gateway_route_table_association
                    aws_ec2_transit_gateway_route_table_propagation
                    aws_ec2_volume_attachment
                    aws_ec2_vpc_peering_connection
                    aws_ec2_vpn_gateway_route_propagation
                    aws_flow_log
                    aws_internet_gateway
                    aws_nat_gateway
                    aws_network_acl
                    aws_route_table
                    aws_security_group
                    aws_subnet
                    aws_transit_gateway
                    aws_transit_gateway_connect
                    aws_transit_gateway_multicast_domain
                    aws_transit_gateway_multicast_domain_association
                    aws_transit_gateway_multicast_group_member
                    aws_transit_gateway_multicast_group_source
                    aws_transit_gateway_route
                    aws_vpc
                    aws_vpc_endpoint
                    aws_vpc_endpoint_connection_notification
                    aws_vpc_endpoint_service
                    aws_vpc_endpoint_service_permission
                    aws_vpn_gateway

### Issues Resolved

List any existing issues this PR resolves, or any Discourse or StackOverflow discussion that's relevant

### Check List
Please fill box or appropriate ([x]) or mark N/A.
- [ ] New functionality includes integration tests/controls
- [ ] New Terraform resources
- [ ] Documentation provided or updated for resources 
- [ ] All Integration Tests pass
- [ ] All Unit Tests pass
- [ ] `rake lint` passes
- [ ] All commits have been signed-off for the Developer Certificate of Origin. See <https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco>

### CODE REVIEW

File: libraries/aws_ec2_instance.rb [hunk 1]
```diff
@@ -125,6 +125,10 @@ def role
     end
   end
 
+  def resource_id
+    @instance[:instance_id]
```
Comment (clintoncwolfe) [inline libraries/aws_ec2_instance.rb:129]: This is an example of an access that will fail if the lookup has failed. If @instance is nil, then the [] access will throw an exception. Try it like this, with a dummy instance ID like i-12345678:

```
control "ec2 instance-check" do
  describe aws_ec2_instance("i-12345678") do
    it { should exist }
  end
end
```

This throws an exception for me with the json reporter:
```
[cwolfe@lodi temp]$ be inspec exec -t aws:// cloudy --no-create-lockfile --reporter json > aws.json
[2022-05-25T13:56:29-04:00] WARN: AWS Service Error encountered running a control with Resource aws_ec2_instance. Error message: The instance ID 'i-12345678' does not exist. You should address this error to ensure your controls are behaving as expected.
bundler: failed to load command: inspec (/Users/cwolfe/.rbenv/versions/3.1.1/lib/ruby/gems/3.1.0/bin/inspec)
libraries/aws_ec2_instance.rb:129:in `resource_id': undefined method `[]' for nil:NilClass (NoMethodError)
	from /Users/cwolfe/sandbox/inspec/inspec-5/lib/inspec/reporters/json.rb:59:in `extract_resource_id'
	from /Users/cwolfe/sandbox/inspec/inspec-5/lib/inspec/reporters/json.rb:50:in `block in profile_results'
```

**Review by clintoncwolfe — CHANGES_REQUESTED** [review]
All the resource_ids need to be tested with a case that includes a "miss" lookup and the JSON reporter. Anything that relies on having a search "hit" will fail and throw an exception. We need to fix them. So anything like this:

```
@instance[:instance_id]
```

needs to be rewritten as `@instance ? @instance[:instance_id] : @display_name` (or use "" if display_name is not available)

File: libraries/aws_vpc.rb [hunk 2]
```diff
@@ -61,6 +62,10 @@ def exists?
     !!vpc && !vpc.empty?
   end
 
+  def resource_id
+    @display_name[:vpc_id]
```
Comment (sa-progress) [inline libraries/aws_vpc.rb:66]: @soumyo13 Please correct resource id.

File: libraries/aws_vpn_gateway.rb [hunk 3]
```diff
@@ -48,6 +48,10 @@ def detached?
     !attached?
   end
 
+  def resource_id
+    @opts[:vpn_gateway_id]
```
Comment (sa-progress) [inline libraries/aws_vpn_gateway.rb:52]: Add nil check here aswell

File: libraries/aws_ec2_traffic_mirror_session.rb [hunk 4]
```diff
@@ -34,6 +34,10 @@ def exists?
     !@traffic_mirror_session.nil? && !@traffic_mirror_session.empty?
   end
 
+  def resource_id
+    @traffic_mirror_session[:traffic_mirror_session_id]
```
Comment (sa-progress) [inline libraries/aws_ec2_traffic_mirror_session.rb:38]: Add nil check here as well.

**Review by sa-progress — CHANGES_REQUESTED** [review]

File: libraries/aws_ec2_client_vpn_authorization_rule.rb [hunk 5]
```diff
@@ -37,6 +37,10 @@ def exists?
     !@res.nil? && !@res.empty?
   end
 
+  def resource_id
+    "#{@display_name}_#{@res[:group_id]}"
```
Comment (soumyo13) [inline libraries/aws_ec2_client_vpn_authorization_rule.rb:41]:
```suggestion
    "#{@display_name}_#{@res ? @res[:group_id] : ''}"
```

**Review by sa-progress — APPROVED** [review]

**Pull request merged**